### ISSUE TITLE
[Value diff]LayerNorm output shows minor numerical discrepancies compared to torch.nn.LayerNorm

### ISSUE THREAD

**Issue by ZiangWu-77** [NVIDIA/TransformerEngine#2084]
**Describe the bug**

The LayerNorm implementation within Transformer Engine produces results that are not numerically identical to the standard torch.nn.LayerNorm. When comparing the outputs on the same input tensor with identical weights and bias, torch.allclose returns False. A small number of elements differ between the two outputs, suggesting a potential minor precision or implementation difference.

**Steps/Code to reproduce bug**
run the following code.
```python
import torch
import transformer_engine.pytorch as te

# --- Parameters from the debugging session ---
hidden_size = 1280
batch_size = 5040
eps = 1e-6
device = 'cuda'
dtype = torch.bfloat16

# --- Setup the input tensor ---
# Create a random input tensor to simulate the real scenario
input_tensor = torch.randn(batch_size, hidden_size, device=device, dtype=dtype)

# --- Initialize Transformer Engine LayerNorm ---
te_layernorm = te.LayerNorm(
    hidden_size,
    eps=eps,
    sequence_parallel=False
).to(device=device, dtype=dtype)

# --- Initialize PyTorch's native LayerNorm ---
torch_layernorm = torch.nn.LayerNorm(
    hidden_size,
    eps=eps
).to(device=device, dtype=dtype)

# --- Synchronize weights and bias for a fair comparison ---
with torch.no_grad():
    torch_layernorm.weight.copy_(te_layernorm.weight)
    torch_layernorm.bias.copy_(te_layernorm.bias)

# --- Get outputs from both implementations ---
te_output = te_layernorm(input_tensor)
torch_output = torch_layernorm(input_tensor)

# --- Compare the results ---
are_close = torch.allclose(te_output, torch_output)
print(f"Outputs are close (torch.allclose): {are_close}")

if not are_close:
    num_total_elements = torch_output.numel()
    num_equal_elements = torch.sum(te_output == torch_output).item()
    num_diff_elements = num_total_elements - num_equal_elements
    
    print(f"Total elements: {num_total_elements}")
    print(f"Number of equal elements: {num_equal_elements}")
    print(f"Number of differing elements: {num_diff_elements}")
    
    # Further check the magnitude of the difference
    abs_diff = torch.abs(te_output - torch_output)
    print(f"Maximum absolute difference: {torch.max(abs_diff).item()}")
```
You can get result like:
```bash
Outputs are close (torch.allclose): False
Total elements: 6451200
Number of equal elements: 6451127
Number of differing elements: 73
Maximum absolute difference: 0.015625
```

**Expected behavior**

The output tensor from transformer_engine.pytorch.LayerNorm should be numerically identical (or at least close enough for torch.allclose to pass with default tolerances) to the output from torch.nn.LayerNorm when given the same input, weights, bias, and epsilon value.

**Environment overview (please complete the following information)**

 - Environment location: [Bare-metal, Docker, Cloud(specify cloud provider - AWS, Azure, GCP, Collab)]
 - Method of Transformer Engine install: [pip install or from source]. Please specify exact commands you used to install.
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**
OS version: Ubuntu 22.04
PyTorch version: 2.5.0+cu121
Python version: 3.12
Transformer Engine version: 2.5.0
CUDA version: 12.1

**Device details**
- GPU A800

**Additional context**
This issue was identified during a debugging session inside the apply_normalization function in transformer_engine/pytorch/module/_common.py. The Pdb log below shows the direct comparison that failed:
```bash
(Pdb) from torch.nn import LayerNorm 
(Pdb) torchnorm = LayerNorm(1536, 1e-6)
(Pdb) torchnorm.weight.copy_(ln_weight)
(Pdb) torchnorm.bias.copy_(ln_bias)
(Pdb) torchnorm = torchnorm.to("cuda")
(Pdb) torchnorm = torchnorm.to(torch.bfloat16)
(Pdb) torch.allclose(torchnorm(inputmat), output[0])
False
(Pdb) torch.sum(torchnorm(inputmat)==output[0])
tensor(6450940, device='cuda:0')
(Pdb) output[0].shape 
torch.Size([5040, 1280])
(Pdb) 5040*1280
6451200
```
The log shows that out of 6,451,200 elements, 260 were not identical.

(Note: The Pdb log shows LayerNorm(1536, ...) being instantiated, while the actual output shape is (5040, 1280). The reproduction code above assumes the correct normalized dimension is 1280, based on the output tensor shape. The core numerical discrepancy is present regardless of this detail.)

This discrepancy negatively impacts my model's fine-tuning, leading to worse results than those achieved with the standard torch.nn.LayerNorm

**Comment by ptrendx** [comment]
Hi @ZiangWu-77. This is an expected behavior and a result of the combination of both small differences in the result coming from the different ordering of the computation and the small precision of the BF16 datatype.

In order to better illustrate it consider a following modification to your script:
 
```python
import torch                                                                     
import transformer_engine.pytorch as te                                          
                                                                                 
# --- Parameters from the debugging session ---                                  
hidden_size = 1280                                                               
batch_size = 5040                                                                
eps = 1e-6                                                                       
device = 'cuda'                                                                  
dtype = torch.bfloat16                                                           
                                                                                 
# --- Setup the input tensor ---
# Create a random input tensor to simulate the real scenario
input_tensor = torch.randn(batch_size, hidden_size, device=device, dtype=dtype)

# --- Initialize Transformer Engine LayerNorm ---
te_layernorm = te.LayerNorm(
    hidden_size,
    eps=eps,
    sequence_parallel=False
).to(device=device, dtype=dtype)

# --- Initialize PyTorch's native LayerNorm ---
torch_layernorm = torch.nn.LayerNorm(
    hidden_size,
    eps=eps
).to(device=device, dtype=dtype)

fp32_layernorm = torch.nn.LayerNorm(
    hidden_size,
    eps=eps
).to(device=device)

# --- Synchronize weights and bias for a fair comparison ---
with torch.no_grad():
    torch_layernorm.weight.copy_(te_layernorm.weight)
    torch_layernorm.bias.copy_(te_layernorm.bias)
    fp32_layernorm.weight.copy_(te_layernorm.weight.float())
    fp32_layernorm.bias.copy_(te_layernorm.bias.float())

# --- Get outputs from both implementations ---
te_output = te_layernorm(input_tensor)
torch_output = torch_layernorm(input_tensor)
fp32_output = fp32_layernorm(input_tensor.float())

# --- Compare the results ---
are_close = torch.allclose(te_output, torch_output)
print(f"Outputs are close (torch.allclose): {are_close}")

if not are_close:
    num_total_elements = torch_output.numel()
    num_equal_elements = torch.sum(te_output == torch_output).item()
    num_diff_elements = num_total_elements - num_equal_elements
    
    print(f"Total elements: {num_total_elements}")
    print(f"Number of equal elements: {num_equal_elements}")
    print(f"Number of differing elements: {num_diff_elements}")
    
    # Further check the magnitude of the difference
    abs_diff = torch.abs(te_output - torch_output)
    print(f"maximum absolute difference: {torch.max(abs_diff).item()}")
    diff_fp32 = torch.abs(fp32_output - torch_output.float())
    print(f"maximum absolute difference (torch fp32 vs bf16): {torch.max(diff_fp32).item()}")
    diff_fp32_te = torch.abs(fp32_output - te_output.float())
    print(f"maximum absolute difference (torch fp32 vs te bf16): {torch.max(diff_fp32_te).item()}")
    
    nonequal_elements_te = te_output[te_output != torch_output].flatten()[0]
    nonequal_elements_torch = torch_output[te_output != torch_output].flatten()[0]
    nonequal_elements_fp32 = fp32_output[te_output != torch_output].flatten()[0]
    print(f"fp32 {nonequal_elements_fp32}, torch bf16 {nonequal_elements_torch.float()}, te bf16 {nonequal_elements_te.float()}")
    print(f"torch bf16 error {nonequal_elements_fp32 - nonequal_elements_torch.float()}, te bf16 error {nonequal_elements_fp32 - nonequal_elements_te.float()}")
```

In this modified version I added the third LayerNorm, also coming from pyTorch, computing in FP32, and compared both the BF16 pyTorch and TE results to it. This is what I get when I run this code:

```bash
Outputs are close (torch.allclose): False                                                                                                                                                                          
Total elements: 6451200                                                                                                                                                                                            
Number of equal elements: 6451139                                                                                                                                                                                  
Number of differing elements: 61                                                                                                                                                                                   
maximum absolute difference: 0.015625                                                                                                                                                                              
maximum absolute difference (torch fp32 vs bf16): 0.015618324279785156                                                                                                                                             
maximum absolute difference (torch fp32 vs te bf16): 0.015618324279785156                                                                                                                                          
fp32 -1.5195313692092896, torch bf16 -1.5234375, te bf16 -1.515625                                                                                                                                                 
torch bf16 error 0.0039061307907104492, te bf16 error -0.003906369209289551   
```

There are 2 things I added to the comparison. First, I compare the max absolute difference between the BF16 results and the FP32 result. As you can see, the error from using BF16 is the same as the error between the 2 BF16 implementations, so it is on the order of BF16 epsilon.
Next, to show why we have the difference between TE and pyTorch in BF16, we also look at one such mismatched value. As you can see, the FP32 value lies basically exactly in the middle between the values given by the BF16 implementations. In such case, which is rare but possible when the number of the elements in the input tensor is large, any small (~1e-7) difference (coming from the ordering of the computation) in the FP32 value computed internally in the kernel could move it closer to one or the other BF16 value and that is the case here.

**Comment by ZiangWu-77** [comment]
Hi @ptrendx That makes sense now. I really appreciate your patient explanation!
